proposals/0005-nio-http-client.md

```diff
@@ -0,0 +1,183 @@
+```swift
+class HTTPClient {
+
+    public func get(url: String, timeout: Timeout? = nil) -> EventLoopFuture<Response> {
+    }
+
+    public func post(url: String, body: Body? = nil, timeout: Timeout? = nil) -> EventLoopFuture<Response> {
+    }
+
+    public func put(url: String, body: Body? = nil, timeout: Timeout? = nil) -> EventLoopFuture<Response> {
+    }
+
+    public func delete(url: String, timeout: Timeout? = nil) -> EventLoopFuture<Response> {
+    }
+
+    public func execute(request: Request, timeout: Timeout? = nil) -> EventLoopFuture<Response> {
+    }
+
+    public func execute<T: HTTPClientResponseDelegate>(request: Request, delegate: T, timeout: Timeout? = nil) -> Task<T.Response> {
+    }
+}
+```
+For the first release we have the following features implemented:
+1. Simple follow-redirects (cookie headers are dropped)
+2. Streaming body download
+3. TLS support
+4. Cookie parsing (but not storage)
+
+## Detailed design
+
+### Lifecycle
+Creating a client is strait-forward:
+```swift
+let client = HTTPClient(eventLoopGroupProvider: .createNew)
+```
+This initializer will create new `EventLoopGroup`, so every client instance will have it's own separate `EventLoopGroup`. Alternatively, users can supply shared `EventLoopGroup`:
+```swift
+let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+let client = HTTPClient(eventLoopGroupProvider: .shared(group))
+```
+It is important to shutdown your client instance after it is no longer needed, in order to shutdown internal `SwiftNIO` machinery:
+```swift
+try client.syncShutdown()
+```
+In case `EventLoopGroup` was provided to the client instance, there is no need to shutdown the client, we expect that lifecycle of that group will be controlled by its owner.
+
+### Request
+In case helper methods do not provide required functionality (for example, if user needs to set headers, or use specific HTTP method), clients of the library can use `HTTPClient.Request`:
+```swift
+extension HTTPClient {
+     enum Body {
+        case byteBuffer(ByteBuffer)
+        case data(Data)
+        case string(String)
+    }
+
+    struct Request {
+        public var version: HTTPVersion
+        public var method: HTTPMethod
+        public var url: URL
+        public var headers: HTTPHeaders
+        public var body: Body?
+
+        public init(url: String,
+                    version: HTTPVersion = HTTPVersion(major: 1, minor: 1),
+                    method: HTTPMethod = .GET,
+                    headers: HTTPHeaders = HTTPHeaders(),
+                    body: Body? = nil) throws {}
+    }
+}
+```
+Example:
+```swift
+var request = try HTTPClient.Request(url: "http://swift.org")
+request.headers.add(name: "User-Agent", value: "nio-http-client")
+
+let future = client.execute(request: request)
+```
+
+### Response
+`HTTPClient`'s methods return an `EventLoopFuture<HTTPClient.Reponse`. This struct is defined as follows:
```

Comment:
Minor typo here, missing `>`, fixed in https://github.com/swift-server/sswg/pull/18